near-sdk/src/store/vec/mod.rs

```diff
@@ -93,6 +147,16 @@ where
     }
 }
 
+#[test]
+fn collections_vec_not_backwards_compatible() {
+    use crate::collections::Vector as Vec1;
+
+    let mut v1 = Vec1::new(b"m");
+    v1.extend([1u8, 2, 3, 4]);
+    // Old collections serializes length as `u64` when new serializes as `u32`.
+    assert!(Vector::<u8>::try_from_slice(&v1.try_to_vec().unwrap()).is_err());
```

Comment:
hmm, yeah. It's definitely opinionated, and was thinking of doing it in a separate PR because of that, but I can do it now.

I'm thinking of just a function like:
```rust
impl Vector<T> {
  fn to_v2(&self) -> crate::store::Vector { ... }
}
```

Reason being is that the data that is cloned is small, and most migration functions don't have owned `self` for some reason (assuming because the SDK marks it as a view call, but unclear why that would matter).

@itegulov do you have an intuition of what the API would look like?

Reply:
This looks reasonable to me. The whole structure is literally just length + prefix, so cloning does not seem to be a big issue to me.

> most migration functions don't have owned self for some reason (assuming because the SDK marks it as a view call, but unclear why that would matter).

We had a conversation about this recently. Functions that take owned self are considered to be view and they can't write state back to storage, so they wouldn't be useful for migration anyway, no?

Unless I am misunderstanding something usually a user would version their contract like this:
```
enum Contract {
    V0(StateV0),
    V1(StateV1),
}
```
Where `StateV0` uses old collections and `StateV1` uses new collections. And then they would add a migration function that migrates all fields from `StateV0` using one of the functions provided by us (`.to_v2()`). Right?

Reply:
That's only if they structured their contract to be versioned from the start. If you don't, you would likely need to run a manual migration, otherwise, you wouldn't know which version of the object exists in storage if you tried to do ambiguously

Reply:
Looks good to me, but are we thinking of getting rid of the old collections eventually? If so, we would have to move this function elsewhere than collections::Vector and someone might forget to do so. So probably best to have it be apart of the new collections as well, like `from_v1`